apps/passport-client/src/validateState.ts

```diff
@@ -217,6 +289,40 @@ export function getRunningAppStateValidationErrors(
     }
   }
 
+  if (
+    commitmentV4FromSelfField === undefined ||
+    commitmentOfIdentityV4PCDInCollection === undefined ||
+    commitmentV4FromIdentityField === undefined
+  ) {
+    // these cases are validated earlier in this function
+  } else {
+    // in 'else' block we check that the commitments from all three
+    // places that the user's commitment exists match - in the self, the
+    // identity, and in the pcd collection
+
+    if (commitmentOfIdentityV4PCDInCollection !== commitmentV4FromSelfField) {
+      errors.push(
+        `commitment of identity v4 pcd in collection (${commitmentOfIdentityV4PCDInCollection})` +
+          ` does not match commitment in 'self' field of app state (${commitmentV4FromSelfField})`
+      );
+    }
+    if (commitmentV4FromSelfField !== commitmentV4FromIdentityField) {
+      errors.push(
+        `v4 commitment in 'self' field of app state (${commitmentV4FromSelfField})` +
+          ` does not match commitment of 'identityV4' field of app state (${commitmentV4FromIdentityField})`
+      );
+    }
+
+    if (
+      commitmentV4FromIdentityField !== commitmentOfIdentityV4PCDInCollection
+    ) {
+      errors.push(
+        `commitment of 'identityV4' field of app state (${commitmentV4FromIdentityField})` +
+          ` does not match commitment of identity v4 pcd in collection (${commitmentOfIdentityV4PCDInCollection})`
+      );
+    }
+  }
+
```

Comment:
Add an explicit check of the migration rule.  I.e. if the user has both v3 and v4 identities, then one should be derived from the other.